client/serviceregistration/checks/client.go

```diff
@@ -164,6 +164,14 @@ func (c *checker) checkHTTP(ctx context.Context, qc *QueryContext, q *Query) *st
 		return qr
 	}
 	request.Header = q.Headers
+
+	for key, values := range q.Headers {
+		if strings.EqualFold(key, "Host") && len(values) > 0 {
+			request.Host = values[0]
+			delete(q.Headers, key)
+		}
+	}
```

Comment:
It's always fun getting to the bottom of stuff...

It seems like sending an HTTP request with a header called `hoST` results in a 400 because of https://github.com/golang/go/issues/18215 .. The spec says a `Host` header is mandatory, but since headers are case insensitive a `hoST` header on a request means it should be valid.

Although I've been unable to confirm this in the specs, many resources (like MDN) state that intermediate parties should pass on headers unmodified. One could argue that if our configuration format supports setting the casing for headers, we should pass on the headers exactly as specified...

I'm continuing on the assumption that in this case we do not care enough and it is okay to canonicalize the headers. Which means that I can also remove the test where a non-canonicalized name is sent (which then fails the test because the Go http server rejects it immediately).

Reply:
More fun in Go! 
It seems the features from `http.Header` are not enforced when using direct access:
https://go.dev/play/p/loMtPja3AUj

So this means that even if we get an object of the type `http.Header` we do not know if the `.Get` and `.Set` api were used or whether direct access was used. Essentially if a class is not closed for modification from the outside its type does not give us any guarantees :-|

Reply:
> Essentially if a class is not closed for modification from the outside its type does not give us any guarantees :-|

Yeah Go's type system leaves a lot to be desired on that front. I think the approach you've got in the most recent commit is the best we're going to be able to do.